sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubPartitionAsyncConsumer.java

```diff
@@ -3,127 +3,68 @@
-        RECEIVE_LINK_FIELD_UPDATER = AtomicReferenceFieldUpdater.newUpdater(
-        EventHubPartitionAsyncConsumer.class, AmqpReceiveLink.class, "receiveLink");
-
-    // We don't want to dump too many credits on the link at once. It's easy enough to ask for more.
-    private static final int MINIMUM_REQUEST = 0;
-    private static final int MAXIMUM_REQUEST = 100;
-
-    private final AtomicInteger creditsToRequest = new AtomicInteger(1);
+    private final ClientLogger logger = new ClientLogger(EventHubPartitionAsyncConsumer.class);
     private final AtomicBoolean isDisposed = new AtomicBoolean();
     private final AtomicReference<LastEnqueuedEventProperties> lastEnqueuedEventProperties = new AtomicReference<>();
-    private final ClientLogger logger = new ClientLogger(EventHubPartitionAsyncConsumer.class);
+    private final AmqpReceiveLinkProcessor amqpReceiveLinkProcessor;
     private final MessageSerializer messageSerializer;
     private final String fullyQualifiedNamespace;
     private final String eventHubName;
     private final String consumerGroup;
     private final String partitionId;
-    private final EmitterProcessor<PartitionEvent> emitterProcessor;
-    private final Flux<PartitionEvent> messageFlux;
     private final boolean trackLastEnqueuedEventProperties;
+    private final EmitterProcessor<PartitionEvent> emitterProcessor;
 
-    private volatile AmqpReceiveLink receiveLink;
-
-    EventHubPartitionAsyncConsumer(Mono<AmqpReceiveLink> receiveLinkMono, MessageSerializer messageSerializer,
-        String fullyQualifiedNamespace, String eventHubName, String consumerGroup, String partitionId,
-        int prefetchCount, boolean trackLastEnqueuedEventProperties) {
+    EventHubPartitionAsyncConsumer(AmqpReceiveLinkProcessor amqpReceiveLinkProcessor,
+        MessageSerializer messageSerializer, String fullyQualifiedNamespace, String eventHubName, String consumerGroup,
+        String partitionId, AtomicReference<EventPosition> currentEventPosition,
+        boolean trackLastEnqueuedEventProperties) {
+        this.amqpReceiveLinkProcessor = amqpReceiveLinkProcessor;
         this.messageSerializer = messageSerializer;
         this.fullyQualifiedNamespace = fullyQualifiedNamespace;
         this.eventHubName = eventHubName;
         this.consumerGroup = consumerGroup;
         this.partitionId = partitionId;
-        this.emitterProcessor = EmitterProcessor.create(prefetchCount, false);
+        this.emitterProcessor = amqpReceiveLinkProcessor
+            .map(message -> onMessageReceived(message))
+            .doOnNext(event -> {
+                // Keep track of the last position so if the link goes down, we don't start from the original location.
+                final Long offset = event.getData().getOffset();
+                if (offset != null) {
+                    currentEventPosition.set(EventPosition.fromOffset(offset));
```

Comment:
Would it be more efficient if we just kept track of offset (long value) and convert to EventPosition only when the link goes down to restore the link? It prevents creating a lot of EventPosition objects.

Reply:
Makes sense. I'll update it.